### ISSUE TITLE
Git tags not resolving to the right commits

### ISSUE THREAD

**Issue by PaulSonOfLars** [argoproj/argo-cd#20083]
**Describe the bug**

Pinning to a git tag uses the wrong commit.

**To Reproduce**

An example app to reproduce this is available here: https://github.com/PaulSonOfLars/test-argo-app

Note that the repo has two tags:
- `2024-09-22-test`, against commit `4650a611850c03600e423d8385ab237308a8ff05`
- `2024-09-22-test-2`, against commit `6650b7ff386c9cb8a2e199ae1fd78571aa3ac336`

The app is pinned to `2024-09-22-test`, so SHOULD be using commit `4650...`. However, it is not - see below;
<img width="640" alt="image" src="https://github.com/user-attachments/assets/9e6aac7c-b7a8-44c1-895f-6d8e28a2c51d">

**Expected behavior**

ArgoCD should use the commit that the tag is pinned to (`4650...`). However, it is using `6650...` instead.

**Screenshots**

See git history below, with commits etc.
<img width="440" alt="image" src="https://github.com/user-attachments/assets/0d572c11-6ab3-4347-9e93-30b9ff087284">

**Version**

`v2.12.3+6b9cd82`

**Logs**

Have enabled debug logs, but can't see anything of note.

My initial reaction is that this may be an issue to do #17566; somehow the tags above are being treated as semver, when they shouldn't be. This then leads to the tags being updated to later commits, even though we expect them to be fixed.
This would explain why ArgoCD uses the second tag, rather than HEAD.

Does that sound about right, or am I barking up the wrong tree?

**Comment by anandf** [comment]
This breaking change has been introduced between `v2.11.0-rc3` and `v2.11.0`. 
I tested this regression with the git repo -> https://github.com/anandf/argocd-example-apps and revision `v0.1.0` which is an annotated tag.
```
argocd app create guestbook --core --repo https://github.com/anandf/argocd-example-apps --path kustomize-guestbook --revision v0.1.0 --dest-namespace guestbook --dest-server https://kubernetes.default.svc --sync-policy automated --self-heal --sync-option CreateNamespace=true
```
Broken in `v2.11.0`
![Screenshot 2024-09-26 at 3 44 23 PM](https://github.com/user-attachments/assets/75389243-0816-4ce5-a742-04a440ef3e5f)
v0.1.0 (256a701) -> is the annotated tag commit id pointing to https://github.com/anandf/argocd-example-apps/commit/256a701b08f64d8191a983a62fa7e5c6dc004e90
Last working version in `v2.11.0-rc3`
![Screenshot 2024-09-26 at 4 35 02 PM](https://github.com/user-attachments/assets/db88ab9f-9623-4faf-805b-1e983a59e2e2)
v0.1.0 (d7927a2) -> is the actual commit id pointing to a valid url https://github.com/anandf/argocd-example-apps/commit/d7927a27b4533926b7d86b5f249cd9ebe7625e90

I am suspecting this commit to have caused this regression. https://github.com/argoproj/argo-cd/commit/da6c2e9c087d7782cd4d35847e676a1b5340f351

**Comment by PaulSonOfLars** [comment]
@aenshin-pp thanks for the reference! Sadly, I don't think this is to do with lightweight/annotated tags - I did come across the PR you've linked while I was investigating, and double checked that I am definitely using lightweight tags for all purposes here.

I can confirm this because if I pin to a tag `2024-09-22-test`, it will correctly target the expected commit. This will remain as expected even if i push new commits.
But as soon as I create a new tag `2024-09-22-test2` (and DONT change the pin), the app will instead still start tracking the new commit referenced by the new tag (even though the app still pins to `2024-09-22-test`)

I opened up #20096 and wrote some tests to confirm my assumptions, and can confirm the issue; argocd seems to be interpreting genuine non-semver git tags as semver constraints, which are then being smudged into using "newer versions".

**Comment by PaulSonOfLars** [comment]
> I am suspecting this commit to have caused this regression. [da6c2e9](https://github.com/argoproj/argo-cd/commit/da6c2e9c087d7782cd4d35847e676a1b5340f351)

That seems surprising, given that commit doesnt seem to have anything to do with tags or git, and argocd didnt have semver resolution at the time - that was a feature introduced in v12...

**Comment by aenshin-pp** [comment]
@PaulSonOfLars got it. Your issue sound more complicated then mine :) 
Mine is here: https://github.com/argoproj/argo-cd/issues/20082 and it is the same as old one https://github.com/argoproj/argo-cd/issues/4231 . It relates annotated tags and easy to reproduce which @anandf has confirmed.

**Comment by blakepettersson** [comment]
@anandf for me it seems to be the other way round 🤔:

v2.11.0-rc3 (`256a701`): 

<img width="764" alt="Screenshot 2024-09-29 at 15 33 50" src="https://github.com/user-attachments/assets/5fee7952-1030-493c-9630-d7b42663ebc8">

v2.11.0 (`d7927a2`):

<img width="760" alt="Screenshot 2024-09-29 at 15 37 01" src="https://github.com/user-attachments/assets/4a275ced-ea5b-496c-a27e-40383c86443f">

**Comment by blakepettersson** [comment]
Oh interesting, when running v2.11.0-rc3 again, I get this:

<img width="762" alt="Screenshot 2024-09-29 at 15 48 53" src="https://github.com/user-attachments/assets/41ec432a-bf74-493b-a132-f590f340c359">

So I guess what this says is that

1. this error predates v2.11.0-rc3
2. this doesn't always happen

**Comment by blakepettersson** [comment]
Hmm, I noticed that with rc3 that it starts with `d7927a2`, but after a while the sync status shows `256a701`

**Comment by PaulSonOfLars** [comment]
@blakepettersson Unsure if this was made clear from the above conversation, but just to avoid any confusion, as it seems youre investigating the wrong repro -
The issues raised by @anandf and @aenshin-pp are related to lightweight vs annotated tags, and are a different issue to the one I'm raising here, which is regarding semver resolution of git tags. 

The repro I've provided in the issue description is only using lightweight tags; adding a new tag starting with the same number will then "bump" to the new commit, while showing the old tag, because tags are being treated as semver constraints rather than an actual tag

**Comment by blakepettersson** [comment]
@PaulSonOfLars I'm aware that this issue is different from yours, I just wanted to give this a quick spin. 

I'll take a look at your issue in a bit - preliminarily my thinking is that your PR should be merged as-is, and then a subsequent PR would try to consolidate the semver resolution between Helm/Git/(and soon OCI).

**Comment by PaulSonOfLars** [comment]
Ok great - thanks for clarifying! Just wanted to make sure :) 

FWIW - I agree with breaking it down into two steps, and have some code ready for that second PR already. Since I'm already familiar with that corner of the code, I figured I might as well!

**Comment by blakepettersson** [comment]
@anandf @aenshin-pp I `git bisect`ed this, and it seems like the actual culprit is #17874, which got introduced in 2.11-rc3.

cc @alexmt

**Comment by ryanisnan** [comment]
We just got burned by this.

**Comment by blakepettersson** [comment]
@ryanisnan by annotated tags or faulty semver resolution? (nb: we should move the annotated tags discussion to #20096 to reduce the confusion)

**Comment by blakepettersson** [comment]
@PaulSonOfLars's PR #20096 fixed this issue. For the other issue we can carry on in #20082.

**Comment by blakepettersson** [comment]
@ryanisnan could you provide some details, e.g your Application spec? Are you using semver resolution?

**Comment by blakepettersson** [comment]
@modulo1982 which issue are you experiencing? There are two issues talked about in this thread - the semver fix will be backported to 2.12.

**Comment by modulo1982** [comment]
We are running into the semver issue we are running into. Good to hear it will also hit 2.12. And also found the pull request for 2.12 so will follow that. Thanks!

**Comment by ryanisnan** [comment]
Hey @blakepettersson , sorry for the delay. We are using tag-based resolution I believe, e.g. `v1`. Is that what you mean?

**Comment by blakepettersson** [comment]
@ryanisnan your issue should be fixed with 2.12.6
